File: src/Roguelike.Core/Core/Data/DataPool/EnemyPool.cs
using System.Collections.Generic;
using System.Linq;
using System;

namespace Roguelike.Data
{
    /// <summary>
    /// A data container that holds all possible EnemyData templates for the game.
    /// Acts as a central database for enemies
    /// </summary>
    public class EnemyPool
    {
        /// <summary>
        /// The primary storage for all enemies, indexed by their unique string ID for fast lookups
        /// </summary>
        public Dictionary<string, EnemyData> EnemiesById { get; set; } = new Dictionary<string, EnemyData>();

        /// <summary>
        /// A secondary index grouping enemies by their star rating for quick access.
        /// </summary>
        public Dictionary<int, List<EnemyData>> EnemiesByStar { get; set; } = new Dictionary<int, List<EnemyData>>();
        
        /// <summary>
        /// Initializes the EnemyPool
        /// </summary>
        public void Initialize(IEnumerable<EnemyData> allEnemies)
        {
            EnemiesById.Clear();
            EnemiesByStar.Clear();

            foreach(var enemy in allEnemies)
            {
                EnemiesById[enemy.Id] = enemy;
                
                if (!EnemiesByStar.ContainsKey(enemy.StarRating))
                {
                    EnemiesByStar[enemy.StarRating] = new List<EnemyData>();
                }
                EnemiesByStar[enemy.StarRating].Add(enemy);
            }
        }
        
        /// <summary>
        /// Retrieves an enemy by its unique ID.
        /// </summary>
        public EnemyData GetEnemy(string id)
        {
            EnemiesById.TryGetValue(id, out var enemy);
            return enemy;
        }

        /// <summary>
        /// Gets a random enemy of a specific Star Rating.
        /// </summary>
        public EnemyData GetRandomEnemyOfStar(int star, Random rng)
        {
            if (EnemiesByStar.TryGetValue(star, out var list) && list.Any())
            {
                return list[rng.Next(list.Count)];
            }
            if (star > 1) return GetRandomEnemyOfStar(star - 1, rng);
            return null;
        }
        
        /// <summary>
        /// Gets a random enemy with Star Rating strictly less than the provided value.
        /// Used for filling out encounters (minions).
        /// </summary>
        public EnemyData GetRandomEnemyBelowStar(int starLimit, System.Random rng)
        {
            var validStars = EnemiesByStar.Keys.Where(k => k < starLimit).ToList();
            
            if (!validStars.Any())
            {
                if (EnemiesByStar.Count > 0)
                {
                    int min = EnemiesByStar.Keys.Min();
                    return GetRandomEnemyOfStar(min, rng);
                }
                return null;
            }

            int chosenStar = validStars[rng.Next(validStars.Count)];
            return GetRandomEnemyOfStar(chosenStar, rng);
        }
    }
}

File: python_roguelike/data/pools/enemy_pool.py
import random
from typing import Dict, List, Optional
from ..combatant.enemy_data import EnemyData


class EnemyPool:
    def __init__(self):
        self.enemies_by_id: Dict[str, EnemyData] = {}
        self.enemies_by_star: Dict[int, List[EnemyData]] = {}

    def initialize(self, all_enemies: List[EnemyData]):
        self.enemies_by_id.clear()
        self.enemies_by_star.clear()

        for enemy in all_enemies:
            self.enemies_by_id[enemy.id] = enemy
            if enemy.star_rating not in self.enemies_by_star:
                self.enemies_by_star[enemy.star_rating] = []
            self.enemies_by_star[enemy.star_rating].append(enemy)

    def get_enemy(self, id: str) -> Optional[EnemyData]:
        return self.enemies_by_id.get(id)

    def get_random_enemy_of_star(self, star: int, rng: random.Random) -> Optional[EnemyData]:
        lst = self.enemies_by_star.get(star)
        if lst:
            return rng.choice(lst)
        if star > 1:
            return self.get_random_enemy_of_star(star - 1, rng)
        return None

    def get_random_enemy_below_star(self, star_limit: int, rng: random.Random) -> Optional[EnemyData]:
        valid_stars = [k for k in self.enemies_by_star if k < star_limit]
        if not valid_stars:
            if self.enemies_by_star:
                min_star = min(self.enemies_by_star.keys())
                return self.get_random_enemy_of_star(min_star, rng)
            return None
        chosen_star = rng.choice(valid_stars)
        return self.get_random_enemy_of_star(chosen_star, rng)
